src/TimeoutMigrationTool/RavenDB/Raven4Adapter.cs

```diff
@@ -321,19 +325,19 @@ public async Task BatchDelete(string[] keys)
 
             while (queryStringIds.Any())
             {
-                try
-                {
-                    var idQry = queryStringIds.First();
-                    uriBuilder.Append($"{idQry}&");
-                    queryStringIds.Remove(idQry);
-                }
-                catch (ArgumentOutOfRangeException)
+                var idQry = queryStringIds.First();
+                var token = $"{idQry}&";
+
+                if (uriBuilder.Length > uriBuilder.MaxCapacity - token.Length)
                 {
                     var uri = uriBuilder.ToString().TrimEnd('&');
                     uris.Add(uri);
                     uriBuilder = new StringBuilder(RavenConstants.MaxUriLength, RavenConstants.MaxUriLength);
                     uriBuilder.Append(url);
                 }
+
+                uriBuilder.Append(token);
+                queryStringIds.Remove(idQry);
             }
```

Comment:
@mauroservienti @lailabougria we had to change this because the behaviour of `StringBuilder.Append` with respect to `MaxCapacity` is not deterministic.

Sometimes, when it throws the `ArgumentOutOfRangeException`, the string is still appended to the `StringBuilder`, and sometimes it is not.

![blinking-meme-gif-7](https://user-images.githubusercontent.com/677704/155732694-4dc709e1-a1fe-4831-88bc-6d3fca7786a5.gif)

From https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.maxcapacity?view=net-6.0

> In .NET Core and in the .NET Framework 4.0 and later versions, when you instantiate the StringBuilder object by calling the StringBuilder(Int32, Int32) constructor, both the length and the capacity of the StringBuilder instance can grow beyond the value of its MaxCapacity property. This can occur particularly when you call the Append(String) and AppendFormat(String, Object) methods to append small strings.

The problem this was causing was that some tests were occasionally getting duplicate documents back from this method, ending up with test failures like [this](https://github.com/Particular/TimeoutMigrationTool/runs/5318909148?check_suite_focus=true#step:15:460):

>    Error: CanReadDocumentsBySpecifiedIdsWhenNumberOfIdsWouldResultInTooLongUri:   Expected: 500
>     Failed CanReadDocumentsBySpecifiedIdsWhenNumberOfIdsWouldResultInTooLongUri [444 ms]
>     Error Message:
>        Expected: 500
>     But was:  501
>   
>     Stack Trace:
>        at TimeoutMigrationTool.Raven.IntegrationTests.RavenAdapterTests.CanReadDocumentsBySpecifiedIdsWhenNumberOfIdsWouldResultInTooLongUri() in /_/src/TimeoutMigrationTool.Raven.IntegrationTests/RavenAdapterTests.cs:line 57

Do you think this could be a problem in production? I.e. someone running the tool may end up with duplicate migrated timeouts? Or is there some other de-duplication happening which would remove any duplicates?

Reply:
we'll investigate this offline

Reply:
Result: It was a bug, but a fairly minor one, that is now accidentally fixed but not released.

https://github.com/Particular/TimeoutMigrationTool/issues/353 will track the issue for inclusion in the next release.